platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -79,21 +80,43 @@ - (void)commonInit
     if (self.URL) {
         NSURL *url = self.URL.mgl_URLByStandardizingScheme;
         source->setURL(url.absoluteString.UTF8String);
-        _features = nil;
+        _feature = nil;
     } else if (self.geoJSONData) {
+        
         NSString *string = [[NSString alloc] initWithData:self.geoJSONData encoding:NSUTF8StringEncoding];
-        const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
-        source->setGeoJSON(geojson);
-        _features = MGLFeaturesFromMBGLFeatures(geojson);
+        // TODO: Figure out a more performant way than serializing the data just to do a type check
+#warning - DEBUG CODE
+        id serialized = [NSJSONSerialization JSONObjectWithData:self.geoJSONData options:0 error:nil];
+        BOOL isFeatureCollection = [serialized isKindOfClass:NSArray.class] || ([serialized isKindOfClass:NSDictionary.class] && [[(NSDictionary *)serialized objectForKey:@"type"] isEqualToString:@"FeatureCollection"]);
+        
+        if (isFeatureCollection) {
+            const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
+            source->setGeoJSON(geojson);
+            _feature = MGLShapeCollectionFeatureFromMBGLFeatures(geojson);
+        } else {
+            const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature>();
+            source->setGeoJSON(geojson);
+            _feature = MGLFeatureFromMBGLFeature(geojson);
+        }
+        
     } else {
-        mbgl::FeatureCollection featureCollection;
-        featureCollection.reserve(self.features.count);
-        for (id <MGLFeaturePrivate> feature in self.features) {
-            featureCollection.push_back([feature mbglFeature]);
+        // Feature collection
+        if ([self.feature isMemberOfClass:MGLShapeCollectionFeature.class]) {
```

Comment:
MGLShapeCollectionFeature implements `-mbglFeature`, so this case is also unnecessary.